src/common/file_buffer.cpp

```diff
@@ -41,30 +46,45 @@ FileBuffer::~FileBuffer() {
 	allocator.FreeData(malloced_buffer, malloced_size);
 }
 
-void FileBuffer::SetMallocedSize(uint64_t &bufsiz) {
-	// make room for the block header (if this is not the db file header)
-	if (type == FileBufferType::MANAGED_BUFFER && bufsiz != Storage::FILE_HEADER_SIZE) {
-		bufsiz += Storage::BLOCK_HEADER_SIZE;
+void FileBuffer::ReallocBuffer(size_t new_size) {
+	if (malloced_buffer) {
+		malloced_buffer = allocator.ReallocateData(malloced_buffer, malloced_size, new_size);
+	} else {
+		malloced_buffer = allocator.AllocateData(new_size);
 	}
-	malloced_size = bufsiz;
-}
-
-void FileBuffer::Construct(uint64_t bufsiz) {
 	if (!malloced_buffer) {
 		throw std::bad_alloc();
 	}
+	malloced_size = new_size;
 	internal_buffer = malloced_buffer;
 	internal_size = malloced_size;
-	buffer = internal_buffer + Storage::BLOCK_HEADER_SIZE;
-	size = internal_size - Storage::BLOCK_HEADER_SIZE;
+	// Caller must update these.
+	buffer = nullptr;
+	size = 0;
 }
 
-void FileBuffer::Resize(uint64_t bufsiz) {
-	D_ASSERT(type == FileBufferType::MANAGED_BUFFER);
-	auto old_size = malloced_size;
-	SetMallocedSize(bufsiz);
-	malloced_buffer = allocator.ReallocateData(malloced_buffer, old_size, malloced_size);
-	Construct(bufsiz);
+void FileBuffer::Resize(uint64_t new_size) {
+	{
+		// TODO: All the logic here is specific to SingleFileBlockManager.
+		// and should be moved there, via a specific implementation of FileBuffer.
+		//
+		// make room for the block header (if this is not the db file header)
+		//
+		// qq: Why is the second condition (new_size != Storage::FILE_HEADER_SIZE) necessary?
```

Comment:
This is likely a special case for the file header (the first 4KB of the file) - we could likely work around it elsewhere if the special case is removed as well.